piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
         } else {
-            let line = self
-                .line_metrics
-                .iter()
-                .find(|l| point.y >= l.y_offset && point.y < l.y_offset + l.height)
-                .unwrap();
-            (true, line)
+            /*
+             * NOTE(ForLoveOfCats): The docs specify that a non-zero value for trailing
+             * indicates that the point aligns to the trailing edge of the grapheme. In
+             * that case the value tells us the number of "characters" in the grapheme.
+             */
+
+            let text = self.pango_layout.get_text().unwrap();
+            let index = index.try_into().unwrap();
+            let trailing = trailing.try_into().unwrap();
+
+            text[index..]
+                .char_indices()
+                .nth(trailing)
+                .map(|(offset, _)| index + offset)
+                .unwrap_or_else(|| text.len())
         };
 
-        // Trailing whitespace is remove for the line
-        let line = &self.text[lm.range()];
-
-        let mut htp = hit_test_line_point(&self.font, line, point);
-        htp.idx += lm.start_offset;
-        if htp.idx == lm.end_offset {
-            htp.idx -= util::trailing_nlf(line).unwrap_or(0);
-        }
-        htp.is_inside &= y_inside;
-        htp
-    }
-
-    fn hit_test_text_position(&self, idx: usize) -> HitTestPosition {
-        let idx = idx.min(self.text.len());
-        assert!(self.text.is_char_boundary(idx));
-
-        if idx == 0 && self.text.is_empty() {
-            return HitTestPosition::new(Point::new(0., self.font.extents().ascent), 0);
-        }
-
-        // first need to find line it's on, and get line start offset
-        let line_num = util::line_number_for_position(&self.line_metrics, idx);
-        let lm = self.line_metrics.get(line_num).cloned().unwrap();
-
-        let y_pos = lm.y_offset + lm.baseline;
+        let (metric_index, metric) = self
+            .line_metrics
+            .iter()
+            .enumerate()
+            .find(|(_, metric)| {
+                metric.start_offset <= index
+                    && index <= metric.end_offset - metric.trailing_whitespace
+            })
+            .unwrap();
 
-        // Then for the line, do text position
-        // Trailing whitespace is removed for the line
-        let line = &self.text[lm.range()];
-        let line_position = idx - lm.start_offset;
+        //NOTE: Manually move to start of next line when hit test is at end of line
+        let index = {
+            let text = self.pango_layout.get_text().unwrap();
+            if index == text.len() || matches!(text.as_bytes()[index], b'\r' | b'\n') {
+                index
+            } else {
+                let mut iterator = GraphemeCursor::new(index, text.len(), true);
+                let next = iterator
+                    .next_boundary(text.as_str(), 0)
+                    .unwrap_or(Some(index))
+                    .unwrap_or(index);
+
```

Comment:
ditto a bit confused here. As I read the docs, the index returned by pango is guaranteed to fall on a grapheme boundary. I'm sure I'm missing something, but I'm not sure what it is.

Reply:
@ForLoveOfCats this is my one remaining question I think!

Reply:
This is just to move the index forward by one grapheme. The goal is that in the case of hitting on the end of a soft break it should move forward to the beginning of the next line in accordance with the behavior that the tests appear to desire.

Reply:
ooooh.

This is about "affinity", then? The position where the soft break occurs has two graphical positions in the layout; the 'upstream' location is the end of the first line, and the 'downstream' location is the start of the second. Druid doesn't currently represent affinity, which means that we always resolve a text position to the 'downstream' location.

This shouldn't involve needing to change the offset, though, because the same offset represents both positions.

In that graphical test, we're using both `hit_test_point` (to get a text position) and then `hit_test_position` to actually find the position in question, in order to draw the cursor. You shouldn't need to change `hit_test_point` to make that work; it should be just a matter of telling `hit_test_position` that you always want the downstream result.

Looking through pango docs, the relevant call seems to be [`index_to_line_x`](https://gtk-rs.org/docs/pango/struct.Layout.html#method.index_to_line_x); if you pass this the index of a boundary, you should get different results based on what argument you pass to `trailing`. I'm not sure how you're supposed to use the result, but I haven't dug too deeply. Maybe it's through the line object, with [`index_to_x`](https://gtk-rs.org/docs/pango/struct.LayoutLine.html#method.index_to_x)?

let me know if any of this doesn't make sense. I would be slightly worried that advancing the cursor here could cause a bug (that would visually look very similar to the bug above, where the hit-tests were off-by-one?)

Reply:
That's not the issue that this is solving. Pango seems to always give us the index of the last character in the line when we hit on or past it which means we get the cursor position just to the left of it. This logic moves that index to the right of that last character. Without it the assertion on line 179 of `piet-common/tests/text.rs` fails with an index of 2 when it expected 3.

I'm not confident about skipping this when the character in question is a hard newline but that is what the tests seem to require such as line 190 of the same file (and same test) where the line in question contains only a lone `\n` and the test expects the index of that newline character.

Reply:
Resolving this conversation as it was misinterpreting the actual problem that this snippet was attempting to solve, more discussion has taken place on Zulip, and the attached piece of the diff is now outdated